src/main/java/com/mraof/minestuck/computer/editmode/EditToolDrag.java

```diff
@@ -0,0 +1,356 @@
+	}
+	
+	/**
+	 * Handles code for the recycle tool on the client-side.
+	 */
+	public static void doRecycleCode(TickEvent.ClientTickEvent event)
+	{
+		Minecraft mc = Minecraft.getInstance();
+		if (mc.player == null || event.phase == TickEvent.Phase.END)
+			return;
+		
+		Player player = mc.player;
+		IEditTools cap = player.getCapability(MSCapabilities.EDIT_TOOLS_CAPABILITY, null).orElse(null);
+		if(cap == null)
+			throw new NullPointerException("EditTool Capability is null on player " + player.getDisplayName().toString() + " on client-side!");
+		
+		if (cap.getToolMode() != null && cap.getToolMode() != IEditTools.ToolMode.RECYCLE)
+			return;
+		
+		boolean isDragging = cap.isEditDragging();
+		boolean isDown = mc.options.keyAttack.isDown();
+		
+		if (isDown)
+		{
+			if(!canEditRecycle(player))
+			{
+				cap.resetDragTools();
+				return;
+			}
+			
+			if (!isDragging)
+			{
+				BlockHitResult blockHit = getPlayerPOVHitResult(player.getLevel(), player);
+				if (blockHit.getType() == BlockHitResult.Type.BLOCK)
+				{
+					cap.setToolMode(IEditTools.ToolMode.RECYCLE);
+					
+					cap.setEditPos1(blockHit.getBlockPos());
+					cap.setEditTrace(blockHit.getLocation(), blockHit.getDirection());
+					cap.setEditReachDistance(Math.sqrt(cap.getEditPos1().distToLowCornerSqr(player.getEyePosition().x, player.getEyePosition().y, player.getEyePosition().z)));
+				}
+			}
+			if (cap.getEditPos1() != null)
+			{
+				cap.setEditPos2(getSelectionEndPoint(player, cap));
+				MSPacketHandler.sendToServer(new EditmodeFillPacket(false, isDown, cap.getEditPos1(), cap.getEditPos2(), cap.getEditTraceHit(), cap.getEditTraceDirection()));
+			}
+		}
+		else if (isDragging)
+		{
+			if (cap.getEditPos1() != null)
+			{
+				MSPacketHandler.sendToServer(new EditmodeFillPacket(false, isDown, cap.getEditPos1(), cap.getEditPos2(), cap.getEditTraceHit(), cap.getEditTraceDirection()));
+			}
+			
+			cap.resetDragTools();
+		}
+		
+		cap.setEditDragging(isDown);
+	}
+	
+	/**
+	 * Determines whether the player can use the recycle tool when left-clicking,
+	 * based on the block that they are looking at.
+	 * @param player The client-side editmode player.
+	 * @return True if you are in editmode and NOT looking directly at a multiblock. Else false.
+	 */
+	public static boolean canEditRecycle(Player player)
+	{
+		return (ClientEditHandler.isActive() && !isMultiblock(player));
+	}
+	
+	/**
+	 * Calculates the second corner of a revise/recycle selection,
+	 * based on whether the player is pointing at a block or not,
+	 * and the distance from the player to the block they first highlighted
+	 * at the start of the selection.
+	 * @param player The client-side editmode player.
+	 * @param cap The player's current EditTools capability.
+	 * @return The BlockPos of the second corner of the revise/recycle selection box.
+	 */
+	private static BlockPos getSelectionEndPoint(Player player, IEditTools cap)
```

Comment:
I wonder if instead of taking the capability, what if it took a `double reachDistance, boolean shouldOffsetBlockHit`? The main benefit is that the function could be used before updating the capability.